File: src/autonomous_nav/src/traverse_goals.cpp
#include "nav_util.h"

int main(int argc, char **argv)
{
    CoordNav cn;
    bool publish = true;
    int i = 0;
    int start, current;
    
    cn.set_goals();

    ros::init(argc, argv, "coord_nav");
    MoveBaseClient ac("move_base", true);
    
    ros::NodeHandle n;
    ros::Publisher coord_pub = n.advertise<move_base_msgs::MoveBaseActionGoal>("move_base/goal", 100);
    
    ros::Subscriber status_sub = n.subscribe("/move_base/result", 100, &CoordNav::status_cb, &cn);
    ros::Subscriber pos_sub = n.subscribe("/move_base/feedback", 100, &CoordNav::pos_cb, &cn);

    while(!ac.waitForServer(ros::Duration(5.0)))
    {
        ROS_INFO("Waiting for the move_base action server to come up");
    }
    
    move_base_msgs::MoveBaseActionGoal coord_msg;

    vector<vector<string>> goals = cn.get_coord();

	string pos_x = goals[i][0];
    string pos_y = goals[i][1];
    string ori_z = goals[i][2];
    string ori_w = goals[i][3];

    ros::Rate rate(20);

    start = time(0);

    while(ros::ok())
    {
        current = time(0);
        cn.setElapsed(current - start);
        if(publish == true){
            coord_msg.goal.target_pose.header.frame_id = "odom";
            coord_msg.goal.target_pose.pose.position.x = std::stod(pos_x);
            coord_msg.goal.target_pose.pose.position.y = std::stod(pos_y);
            coord_msg.goal.target_pose.pose.orientation.z = std::stod(ori_z);
            coord_msg.goal.target_pose.pose.orientation.w = std::stod(ori_w);

            coord_pub.publish(coord_msg);
            publish = false;
            ROS_INFO("Received new goal.");
        }
        else if(cn.getStatus() >= 3)
        {
            ROS_INFO("Saved coordinates successfully.");
            cn.save_coord();
            cn.setStatus(0);
            i++;
            if(i < goals.size())
            {
                pos_x = goals[i][0];
                pos_y = goals[i][1];
                ori_z = goals[i][2];
                ori_w = goals[i][3];
                publish = true;
            }
        }
        rate.sleep();
        ros::spinOnce();
    }

    return 0;
}

File: src/autonomous_nav/src/traverse_goals.py
#!/usr/bin/env python

import rospy
import coord_nav
import actionlib
import time
from move_base_msgs.msg import MoveBaseAction, MoveBaseActionGoal, MoveBaseActionResult, MoveBaseActionFeedback


def main():
    cn = coord_nav.CoordNav()
    publish = True
    i = 0
    cn.set_goals()

    rospy.init_node('coord_nav', anonymous=True)
    coord_pub = rospy.Publisher('move_base/goal', MoveBaseActionGoal, queue_size=100)
    rospy.Subscriber('move_base/result', MoveBaseActionResult, cn.status_cb)
    rospy.Subscriber('move_base/feedback', MoveBaseActionFeedback, cn.pos_cb)
   
    client = actionlib.SimpleActionClient('move_base', MoveBaseAction)
    client.wait_for_server()

    coord_msg = MoveBaseActionGoal()
    goal = coord_msg.goal.target_pose
    goals = cn.get_coord()
    pos_x, pos_y, ori_z, ori_w = goals[i]
    
    rate = rospy.Rate(20)

    start_time = time.time()

    while not rospy.is_shutdown():
        current_time = time.time()
        cn.elapsed_time = current_time - start_time
        if publish:
            goal.header.frame_id = 'odom'
            goal.pose.position.x = pos_x
            goal.pose.position.y = pos_y
            goal.pose.orientation.z = ori_z
            goal.pose.orientation.w = ori_w
            coord_pub.publish(coord_msg)
            publish = False
            rospy.loginfo("Received new goal.")
            

        elif cn.status >= 3:
            rospy.loginfo("Saved coordinates successfully.")
            rospy.loginfo(cn.text)
            cn.save_coord()
            cn.status = 0
            i += 1
            if i < len(goals):
                pos_x, pos_y, ori_z, ori_w = goals[i]
                publish = True
            
        rate.sleep()


if __name__ == '__main__':
    try:
        main()
    except rospy.ROSInterruptException: 
        pass